spatialos-sdk/src/worker/schema.rs

```diff
@@ -0,0 +1,582 @@
+    Schema_GetFixed32,
+    Schema_IndexFixed32,
+    Schema_GetFixed32Count,
+    Schema_AddFixed32,
+    Schema_AddFixed32List,
+    Schema_GetFixed32List,
+);
+impl_primitive_field!(
+    u64,
+    SchemaFixed64,
+    Schema_GetFixed64,
+    Schema_IndexFixed64,
+    Schema_GetFixed64Count,
+    Schema_AddFixed64,
+    Schema_AddFixed64List,
+    Schema_GetFixed64List,
+);
+impl_primitive_field!(
+    i32,
+    SchemaSfixed32,
+    Schema_GetSfixed32,
+    Schema_IndexSfixed32,
+    Schema_GetSfixed32Count,
+    Schema_AddSfixed32,
+    Schema_AddSfixed32List,
+    Schema_GetSfixed32List,
+);
+impl_primitive_field!(
+    i64,
+    SchemaSfixed64,
+    Schema_GetSfixed64,
+    Schema_IndexSfixed64,
+    Schema_GetSfixed64Count,
+    Schema_AddSfixed64,
+    Schema_AddSfixed64List,
+    Schema_GetSfixed64List,
+);
+impl_primitive_field!(
+    u32,
+    SchemaEnum,
+    Schema_GetEnum,
+    Schema_IndexEnum,
+    Schema_GetEnumCount,
+    Schema_AddEnum,
+    Schema_AddEnumList,
+    Schema_GetEnumList,
+);
+
+impl SchemaType for EntityId {
+    type RustType = Self;
+
+    fn from_field(object: &SchemaObject, field: FieldId) -> Self::RustType {
+        let id = unsafe { Schema_GetEntityId(object.internal, field) };
+        Self { id }
+    }
+}
+
+impl SchemaIndexType for EntityId {
+    fn field_count(object: &SchemaObject, field: FieldId) -> u32 {
+        unsafe { Schema_GetEntityIdCount(object.internal, field) }
+    }
+
+    fn index_field(object: &SchemaObject, field: FieldId, index: u32) -> Self::RustType {
+        let id = unsafe { Schema_IndexEntityId(object.internal, field, index) };
+        Self { id }
+    }
+}
+
+impl SchemaListType for EntityId {
+    fn get_field_list(object: &SchemaObject, field: FieldId, data: &mut Vec<Self::RustType>) {
+        let count = Self::field_count(object, field) as usize;
+
+        // Ensure that there is enough capacity for the elements in the schema field.
+        if data.capacity() < count {
+            data.reserve(count - data.capacity());
+        }
+
+        // Replace the contents of `data` with the list of values in the schema field.
+        unsafe {
+            data.set_len(count);
+            Schema_GetEntityIdList(object.internal, field, data.as_mut_ptr() as *mut _);
+        }
+    }
+}
+
+impl SchemaType for bool {
+    type RustType = Self;
+
+    fn from_field(object: &SchemaObject, field: FieldId) -> Self::RustType {
+        let raw = unsafe { Schema_GetBool(object.internal, field) };
+        raw != 0
+    }
+}
+
+impl SchemaIndexType for bool {
+    fn field_count(object: &SchemaObject, field: FieldId) -> u32 {
+        unsafe { Schema_GetBoolCount(object.internal, field) }
+    }
+
+    fn index_field(object: &SchemaObject, field: FieldId, index: u32) -> Self::RustType {
+        let raw = unsafe { Schema_IndexBool(object.internal, field, index) };
+        raw != 0
+    }
+}
+
+// TODO: This is only valid if schema bools are guaranteed to be 0 or 1. If schema
```

Comment:
I assume this is referring to the types defined above? Not the `Option<T>`? 

And I believe this is the case. If we receive a value outside of `{ 0, 1 }` it should probably cause a panic

Reply:
First off: Woops, I left that comment on the wrong impl block. That comment is meant to be applied to `impl SchemaListType for bool`, which somehow got lost 😓 

As for why that would be unsound: The Rust `bool` type is represented as an 8 bit unsigned integer with the value 0 or 1. As such, it's [undefined behavior](https://doc.rust-lang.org/beta/reference/behavior-considered-undefined.html) for a `bool` to have any other value. So if a bool in a schema object can have a value other than 0 or 1, we have can't directly transmute it to a bool, we have to manually convert it to `bool` (which is what the `SchemaType` and `SchemaIndexType` impls for `bool` currently do). The problem with implementing `SchemaListType` for `bool` is that we don't have a chance to do that conversion, we're directly populating a `Vec<bool>` from the schema object's data, so we need to guarantee that doing so will only produce valid values. If the schema API semantics don't guarantee that bools will only ever be 0 or 1, then we can't implement `SchemaListType` for `bool`.